crypto/bio/bio_lib.c

```diff
@@ -784,3 +784,100 @@ void bio_cleanup(void)
+
+#ifndef OPENSSL_NO_SOCK
+    if (BIO_get_fd(bio, &fd) > 0)
+        return BIO_socket_wait(fd, BIO_should_read(bio), max_time);
+#endif
+    if (milliseconds > 1000) {
+        long sec_diff = (long)(max_time - time(NULL)); /* might overflow */
+
+        if (sec_diff <= 0)
+            return 0; /* timeout */
+        if ((unsigned long)sec_diff < milliseconds / 1000)
+            milliseconds = (unsigned long)sec_diff * 1000;
+    }
+    ossl_sleep(milliseconds);
+    return 1;
+}
+
+/*
+ * Wait on (typically socket-based) BIO at most until max_time.
+ * Succeed immediately if max_time == 0. If sockets are not available succeed
+ * after waiting at most given milliseconds in order to avoid a tight busy loop.
+ * Call BIOerr(...) unless success.
+ * Returns -1 on error, 0 on timeout, and 1 on success.
+ */
+int BIO_wait(BIO *bio, time_t max_time, unsigned int milliseconds)
+{
+    int rv = bio_wait(bio, max_time, milliseconds);
+
+    if (rv <= 0)
+        BIOerr(0, rv == 0 ? BIO_R_TRANSFER_TIMEOUT : BIO_R_TRANSFER_ERROR);
+    return rv;
+}
+
+/*
+ * Connect via given BIO using BIO_do_handshake() until success/timeout/error.
+ * Parameter timeout == 0 means infinite, < 0 leads to immediate timeout error.
+ * Returns -1 on error, 0 on timeout, and 1 on success.
+ */
+int BIO_connect_retry(BIO *bio, int timeout)
+{
+    int blocking = timeout == 0;
+    time_t max_time = timeout > 0 ? time(NULL) + timeout : 0;
+    int rv;
+
+    if (bio == NULL) {
+        BIOerr(0, ERR_R_PASSED_NULL_PARAMETER);
+        return -1;
+    }
+
+    if (timeout < 0) {
+        BIOerr(0, BIO_R_CONNECT_TIMEOUT);
+        return 0;
+    }
+
+    if (!blocking)
+        BIO_set_nbio(bio, 1);
+
+ retry: /* it does not help here to set SSL_MODE_AUTO_RETRY */
+    rv = BIO_do_handshake(bio); /* This indirectly calls ERR_clear_error(); */
+
+    if (rv <= 0) {
+        if (get_last_sys_error() == ETIMEDOUT) {
+            /*
+             * if blocking, despite blocking BIO, BIO_do_handshake() timed out
+             * when non-blocking, BIO_do_handshake() timed out early
+             * with rv == -1 and get_last_sys_error() == 0
+             */
+            ERR_clear_error();
+            (void)BIO_reset(bio);
+            /*
+             * unless using BIO_reset(), blocking next connect() may crash and
+             * non-blocking next BIO_do_handshake() will fail
+             */
+            goto retry;
+        } else if (BIO_should_retry(bio)) {
+            /* will not actually wait if timeout == 0 (i.e., blocking BIO) */
+            rv = bio_wait(bio, max_time, 100 /* milliseconds */);
+            if (rv > 0)
+                goto retry;
+            BIOerr(0, rv == 0 ? BIO_R_CONNECT_TIMEOUT : BIO_R_CONNECT_ERROR);
+        } else {
+            rv = -1;
+            if (ERR_peek_error() == 0) /* missing error queue entry */
+                BIOerr(0, BIO_R_CONNECT_ERROR); /* workaround: general error */
+        }
+    }
+
+    return rv;
+}
```

Comment:
`BIO_do_handshake()` is a macro closely related to `BIO_f_ssl()`, which means that `BIO_connect_retry()` is by extension.

Looking at previous patterns, we usually implement functions that are closely related to some BIO filter near the implementation of that filter.  In this case, that would be `ssl/bio_ssl.c`.  I would rather see this appear there.

Reply:
`BIO_wait()` is more intuitive to have here, though...  what I would do to simplify matters is not having it record an error, but let the caller do it instead.  I agree that it seems cumbersome, but it will also end up being cumbersome to maintain an externally available internal function, so as trade-offs go...

Reply:
I have the impression that the appearance of `BIO_do_handshake()` in `BIO_connect_retry()` is a bit misleading, and maybe I should revert that. Originally I had been using there `BIO_do_connect(), which is defined as `BIO_do_handshake()`, but the former for some reason is under `# ifndef OPENSSL_NO_SOCK`, so since moving `BIO_connect_retry()` out of that guard it cannot be used there, but  `BIO_do_handshake()` can.
I can imagine that `BIO_connect_retry()` does not fill well in `bio_lib.c`, but I don't think it fits in `bio_ssl.c` because the function is usable not only for SSL BIOs. Where else could I move it?

Reply:
Concerning where to throw the potential errors of `BIO_wait()`, it is not only a matter of convenience to do it within the function, but also a matter of clarity to the consumers of the error info: the error happened at the BIO level. Therefore I prefer to keep the error generation where it is now. I don't think it hurts to have a separate internal static function - I've seen a similar separation also in many other parts of the lib.

Reply:
How about moving the definition of `BIO_do_connect()` out of `#ifndef OPENSSL_NO_SOCK`?

Reply:
I fear I may have had a too narrow view of this.  See, in *our* code, `BIO_do_handshake()` terminates in `crypto/bio/bss_conn.c` or `crypto/bio/bss_acpt.c`, and I failed to consider the unknown, BIOs in the wild that use entirely different mechanisms that we know nothing about.

And looking at this function, it's completely agnostic of the underlying BIO implementation, so it is more fitting here than I was thinking earlier.

In summary: never mind :wink: